### PR TITLE
python3Packages.scikit-learn: unbreak darwin

### PR DESCRIPTION
**Pull request by DavHau**
ZHF: #122042

continuation of https://github.com/NixOS/nixpkgs/pull/122687

###### Things done
disable some failing tests which fail on darwin.

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Added a release notes entry if the change is major or breaking
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 124255` at 4af47234 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>23 packages marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.cnvkit</li>
    <li>python38Packages.dask-ml</li>
    <li>python38Packages.dask-xgboost</li>
    <li>python38Packages.dftfit</li>
    <li>python38Packages.ignite</li>
    <li>python38Packages.mlxtend</li>
    <li>python38Packages.optuna</li>
    <li>python38Packages.osmnx</li>
    <li>python38Packages.pygbm</li>
    <li>python38Packages.shap</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li><a href="https://gist.github.com/c54d961196909293e953d722e9c03afa">python38Packages.scikit-bio</a></li>
    <li><a href="https://gist.github.com/51e3fba2fba697c2701a33e101e9c1ac">python39Packages.qiskit-ignis</a></li>
    <li><a href="https://gist.github.com/a5f34d9309ab7ec39f008b511c235ae3">python39Packages.scikit-bio</a></li>
  </ul>
</details>
<details>
  <summary>22 packages skipped due to time constraints:</summary>
  <ul>
    <li>python38Packages.dask-glm</li>
    <li>python38Packages.graspologic</li>
    <li>python38Packages.hdbscan</li>
    <li>python38Packages.hyppo</li>
    <li>python38Packages.nilearn</li>
    <li>python38Packages.pytorch-metric-learning</li>
    <li>python38Packages.qiskit</li>
    <li>python38Packages.qiskit-aqua</li>
    <li>python38Packages.qiskit-ignis</li>
    <li>python38Packages.scikit-optimize</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>28 packages built successfully:</summary>
  <ul>
    <li>python38Packages.batchgenerators</li>
    <li>python38Packages.bayesian-optimization</li>
    <li>python38Packages.hmmlearn</li>
    <li>python38Packages.imbalanced-learn</li>
    <li>python38Packages.kmapper</li>
    <li>python38Packages.librosa</li>
    <li>python38Packages.lightgbm</li>
    <li>python38Packages.mlrose</li>
    <li>python38Packages.persim</li>
    <li>python38Packages.ppscore</li>
    <li>python38Packages.pynndescent</li>
    <li>python38Packages.ripser</li>
    <li>python38Packages.scikitlearn (python38Packages.scikit-learn)</li>
    <li>python38Packages.seqeval</li>
    <li>python38Packages.word2vec</li>
    <li>python39Packages.batchgenerators</li>
    <li>python39Packages.bayesian-optimization</li>
    <li>python39Packages.hmmlearn</li>
    <li>python39Packages.imbalanced-learn</li>
    <li>python39Packages.kmapper</li>
    <li>python39Packages.lightgbm</li>
    <li>python39Packages.mlrose</li>
    <li>python39Packages.persim</li>
    <li>python39Packages.ppscore</li>
    <li>python39Packages.ripser</li>
    <li>python39Packages.scikitlearn (python39Packages.scikit-learn)</li>
    <li>python39Packages.seqeval</li>
    <li>python39Packages.word2vec</li>
  </ul>
</details>

Note that build failures [may predate](https://github.com/nix-community/hydra-check) this PR, and could be nondeterministic or hardware dependent.
Please exercise your independent judgement. Does something look off? Please file an [issue](https://github.com/rmcgibbo/r-rmcgibbo/issues) or reach out on IRC.

**Review by rmcgibbo — APPROVED** [review]
Result of `nixpkgs-review pr 124255` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>30 packages marked as broken and skipped:</summary>
  <ul>
    <li>python38Packages.cnvkit</li>
    <li>python38Packages.dask-xgboost</li>
    <li>python38Packages.dftfit</li>
    <li>python38Packages.ignite</li>
    <li>python38Packages.mlxtend</li>
    <li>python38Packages.optuna</li>
    <li>python38Packages.osmnx</li>
    <li>python38Packages.pygbm</li>
    <li>python38Packages.pytorch-metric-learning</li>
    <li>python38Packages.qiskit</li>
    <li>python38Packages.qiskit-aqua</li>
    <li>python38Packages.qiskit-ignis</li>
    <li>python38Packages.shap</li>
    <li>python38Packages.sklearn-deap</li>
    <li>python38Packages.skorch</li>
    <li>python38Packages.textacy</li>
    <li>python39Packages.cnvkit</li>
    <li>python39Packages.dask-xgboost</li>
    <li>python39Packages.ignite</li>
    <li>python39Packages.mlxtend</li>
    <li>python39Packages.optuna</li>
    <li>python39Packages.osmnx</li>
    <li>python39Packages.pytorch-metric-learning</li>
    <li>python39Packages.qiskit</li>
    <li>python39Packages.qiskit-aqua</li>
    <li>python39Packages.qiskit-ignis</li>
    <li>python39Packages.sklearn-deap</li>
    <li>python39Packages.skorch</li>
    <li>tebreak</li>
    <li>tts</li>
  </ul>
</details>
<details>
  <summary>14 packages failed to build:</summary>
  <ul>
    <li>python38Packages.graspologic</li>
    <li>python38Packages.hyppo</li>
    <li>python38Packages.lightgbm</li>
    <li>python38Packages.nilearn</li>
    <li>python38Packages.scikit-optimize</li>
    <li>python38Packages.scikit-tda</li>
    <li>python38Packages.umap-learn</li>
    <li>python38Packages.word2vec</li>
    <li>python38Packages.xgboost</li>
    <li>python39Packages.lightgbm</li>
    <li>python39Packages.nilearn</li>
    <li>python39Packages.scikit-optimize</li>
    <li>python39Packages.word2vec</li>
    <li>python39Packages.xgboost</li>
  </ul>
</details>
<details>
  <summary>33 packages built:</summary>
  <ul>
    <li>python38Packages.batchgenerators</li>
    <li>python38Packages.bayesian-optimization</li>
    <li>python38Packages.dask-glm</li>
    <li>python38Packages.dask-ml</li>
    <li>python38Packages.hdbscan</li>
    <li>python38Packages.hmmlearn</li>
    <li>python38Packages.imbalanced-learn</li>
    <li>python38Packages.kmapper</li>
    <li>python38Packages.librosa</li>
    <li>python38Packages.mlrose</li>
    <li>python38Packages.persim</li>
    <li>python38Packages.ppscore</li>
    <li>python38Packages.pynndescent</li>
    <li>python38Packages.ripser</li>
    <li>python38Packages.scikit-bio</li>
    <li>python38Packages.scikitlearn (python38Packages.scikit-learn)</li>
    <li>python38Packages.seqeval</li>
    <li>python38Packages.vowpalwabbit</li>
    <li>python39Packages.batchgenerators</li>
    <li>python39Packages.bayesian-optimization</li>
    <li>python39Packages.dask-glm</li>
    <li>python39Packages.hdbscan</li>
    <li>python39Packages.hmmlearn</li>
    <li>python39Packages.imbalanced-learn</li>
    <li>python39Packages.kmapper</li>
    <li>python39Packages.mlrose</li>
    <li>python39Packages.persim</li>
    <li>python39Packages.ppscore</li>
    <li>python39Packages.ripser</li>
    <li>python39Packages.scikit-bio</li>
    <li>python39Packages.scikitlearn (python39Packages.scikit-learn)</li>
    <li>python39Packages.seqeval</li>
    <li>python39Packages.vowpalwabbit</li>
  </ul>
</details>

**Comment by risicle** [comment]
Package itself certainly builds for me now, macos 10.15, sandbox enabled.

**Review by risicle — APPROVED** [review]
Same x86_64 linux too

**Comment by rmcgibbo** [comment]
I'd love to get this one merged, and then we can see if hydra likes it (fingers crossed!) and fix the downstream builds that are broken before the release. Thanks so much to @nbren12  and @DavHau for all the hard work -- this was not an easy PR.

File: pkgs/development/python-modules/scikit-learn/default.nix [hunk 1]
```diff
@@ -66,10 +66,30 @@ buildPythonPackage rec {
...
+    # verbose build outputs needed to debug hard-to-reproduce hydra failures
+    "-v"
+    "--pyargs" "sklearn"
+
+    # NuSVC memmap tests causes segmentation faults in certain environments
+    # (e.g. Hydra Darwin machines) related to a long-standing joblib issue
+    # (https://github.com/joblib/joblib/issues/563). See also:
+    # https://github.com/scikit-learn/scikit-learn/issues/17582
+    # Since we are overriding '-k' we need to include the 'disabledTests' from above manually.
+    "-k" "'not (NuSVC and memmap) ${toString (lib.forEach disabledTests (t: "and not ${t}"))}'"
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/scikit-learn/default.nix:89]: Why can't we add "NuSVC" "memmap" to disabledTests?

Comment (rmcgibbo) [reply pkgs/development/python-modules/scikit-learn/default.nix:89]: We tried that. It's because disabledTests concatenates the elements with "or", I think. This is sort of matching multiple tests that each contain both of these two sub-strings.

**Comment by nbren12** [comment]
Thanks for taking this over @DavHau! Looking forward to the new release.
